File: clients/Java/1.0 mcp_login/MCPLoginClient.java
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class MCPLoginClient {

    public static void main(String[] args) {
        String serverIp = null;
        int serverPort = 0;
        String email = null;
        String password = null;

        // Argumente parsen
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--server-ip":
                    serverIp = args[++i];
                    break;
                case "--server-port":
                    serverPort = Integer.parseInt(args[++i]);
                    break;
                case "--email":
                    email = args[++i];
                    break;
                case "--password":
                    password = args[++i];
                    break;
            }
        }

        // Validierung
        if (serverIp == null || serverPort == 0 || email == null || password == null) {
            System.out.println("❌ ERROR: Missing required parameters.");
            return;
        }

        System.out.println("🔐 Logging in...");

        // JSON-Objekt erstellen
        JSONObject payload = new JSONObject();
        payload.put("command", "login");

        JSONObject arguments = new JSONObject();
        arguments.put("email", email);
        arguments.put("password", password);
        payload.put("arguments", arguments);

        // Request senden und Antwort empfangen
        String response = sendRequest(serverIp, serverPort, payload);
        System.out.println("✅ Server Response:");
        System.out.println(response);
    }

    /**
     * Stellt eine TCP-Verbindung her, sendet das JSON und empfängt die Antwort.
     */
    public static String sendRequest(String serverIp, int serverPort, JSONObject payload) {
        try (Socket socket = new Socket(serverIp, serverPort)) {
            // JSON in Byte-Array umwandeln
            String payloadString = payload.toString();
            byte[] data = payloadString.getBytes(StandardCharsets.UTF_8);

            // Daten senden
            OutputStream out = socket.getOutputStream();
            out.write(data);
            out.flush();

            // Antwort empfangen
            InputStream in = socket.getInputStream();
            byte[] buffer = new byte[4096];
            int bytesRead = in.read(buffer);

            if (bytesRead == -1) {
                // Falls keine Daten empfangen wurden
                return "❌ ERROR: No response from server.";
            }

            return new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);

        } catch (IOException e) {
            e.printStackTrace();
            return "❌ ERROR: " + e.getMessage();
        }
    }
}

File: clients/JavaScript/1.0 mcp_login/MCPLoginClient.js
const net = require('net');
const readline = require('readline');
const { argv, exit } = require('process');

// Funktion zum Parsen der Kommandozeilenargumente
function parseArguments(args) {
    const parsedArgs = {};
    for (let i = 2; i < args.length; i++) {
        switch (args[i]) {
            case '--server-ip':
                parsedArgs.serverIp = args[++i];
                break;
            case '--server-port':
                parsedArgs.serverPort = parseInt(args[++i], 10);
                break;
            case '--email':
                parsedArgs.email = args[++i];
                break;
            case '--password':
                parsedArgs.password = args[++i];
                break;
            default:
                console.warn(`Unbekanntes Argument: ${args[i]}`);
        }
    }
    return parsedArgs;
}

// Funktion zum Senden einer Anfrage über eine TCP-Verbindung
function sendRequest(serverIp, serverPort, payload) {
    return new Promise((resolve, reject) => {
        const client = new net.Socket();
        let responseData = '';

        client.connect(serverPort, serverIp, () => {
            console.log('🔗 Verbindung zum Server hergestellt.');
            client.write(JSON.stringify(payload));
        });

        client.on('data', (data) => {
            responseData += data.toString();
            // Versuche, die empfangenen Daten als JSON zu parsen
            try {
                const parsedData = JSON.parse(responseData);
                resolve(parsedData);
                client.destroy(); // Verbindung schließen
            } catch (err) {
                // Antwort noch nicht vollständig, weiter empfangen
            }
        });

        client.on('close', () => {
            console.log('🔒 Verbindung zum Server geschlossen.');
        });

        client.on('error', (err) => {
            reject(err);
        });
    });
}

// Funktion zum interaktiven Abfragen des Passworts (optional)
function askPassword(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });

    return new Promise((resolve) => {
        rl.question(query, (password) => {
            rl.close();
            resolve(password);
        });
    });
}

// Hauptfunktion
async function main() {
    const args = argv;
    const { serverIp, serverPort, email, password } = parseArguments(args);

    // Überprüfen, ob alle erforderlichen Parameter vorhanden sind
    if (!serverIp || !serverPort || !email || !password) {
        console.error('❌ ERROR: Fehlende erforderliche Parameter.');
        console.log('Verwendung: node MCPLoginClient.js --server-ip <IP> --server-port <Port> --email <Email> --password <Passwort>');
        exit(1);
    }

    const payload = {
        command: "login",
        arguments: {
            email: email,
            password: password
        }
    };

    try {
        console.log('🔐 Logging in...');
        const response = await sendRequest(serverIp, serverPort, payload);
        console.log('✅ Server Response:');
        console.log(JSON.stringify(response, null, 4));
    } catch (err) {
        console.error('❌ ERROR:', err.message);
    }
}

main();
